src/interpreter/vm.rs

```diff
@@ -63,6 +64,97 @@ pub struct BlockMessages {
     pub win_count: i64,
 }
 
+#[derive(PartialEq, Clone, Debug)]
+pub struct MessageGasCost {
+    pub message: Cid,
+    pub gas_used: BigInt,
+    pub base_fee_burn: TokenAmount,
+    pub over_estimation_burn: TokenAmount,
+    pub miner_penalty: TokenAmount,
+    pub miner_tip: TokenAmount,
+    pub refund: TokenAmount,
+    pub total_cost: TokenAmount,
+}
+
+impl MessageGasCost {
+    pub fn new(msg: &Message, ret: &ApplyRet) -> Self {
+        use crate::message::Message as MessageTrait;
+        Self {
+            message: msg.cid().unwrap(),
+            gas_used: BigInt::from(ret.msg_receipt().gas_used()),
+            base_fee_burn: ret.base_fee_burn(),
+            over_estimation_burn: ret.over_estimation_burn(),
+            miner_penalty: ret.penalty(),
+            miner_tip: ret.miner_tip(),
+            refund: ret.refund(),
+            total_cost: msg.required_funds() - &ret.refund(),
+        }
+    }
+}
+
+#[derive(PartialEq, Clone, Debug)]
+pub struct InvocResult {
+    pub msg_cid: Cid,
+    pub msg: Message,
+    pub msg_receipt: Receipt,
+    pub gas_cost: MessageGasCost,
+    pub execution_trace: Option<Trace>,
+    pub error: String,
+    pub duration: u64,
+}
+
+impl InvocResult {
+    pub fn new(msg_cid: Cid, msg: &Message, ret: &ApplyRet) -> Self {
+        let trace = build_exec_trace(ret.exec_trace());
+        Self {
+            msg_cid,
+            msg: msg.clone(),
+            msg_receipt: ret.msg_receipt(),
+            gas_cost: MessageGasCost::new(msg, ret),
+            execution_trace: trace,
+            error: ret.failure_info().unwrap_or_default(),
+            duration: 0,
+        }
+    }
+}
+
+fn build_exec_trace(exec_trace: Vec<ExecutionEvent_v3>) -> Option<Trace> {
```

Comment:
What does this function do? Maybe give some examples.

Reply:
Definitely needs some sort of doc.

Reply:
@elmattic feel like adding this? I still don't really know what this function does - I just refactored the json stuff

Reply:
The function code has been extracted from this one https://github.com/filecoin-project/filecoin-ffi/blob/v1.23.0/rust/src/fvm/machine.rs#L207.
We're only interested in the part that constructs the trace.

The trace is basically a function call graph and for each function we log all the the operations burning gas. The `build_exec_trace` is converting the `ExecutionEvent` vector to this graph structure, nothing more.